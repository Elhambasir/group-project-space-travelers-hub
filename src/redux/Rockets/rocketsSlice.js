import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const url = 'https://api.spacexdata.com/v3/rockets';

export const getAPI = createAsyncThunk(

  'Rockets/fetchRocket', async () => {
    try {
      const response = await axios.get(url);
      return response.data;
    } catch (err) {
      return err.message;
    }
  },
);

const initialState = {
  rocketList: [],
  isLoading: false,
  error: undefined,
};

const rocketsSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
  },
  extraReducers(builder) {
    builder
      .addCase(getAPI.pending, (state) => {
        state.isLoading = true;
      })
      .addCase(getAPI.fulfilled, (state, action) => {
        const rockets = [];
        action.payload.map((item) => {
          let newRocket = {};
          newRocket = {
            id: item.id,
            name: item.rocket_name,
            type: item.rocket_type,
            images: item.flickr_images,
          };
          return rockets.push(newRocket);
        });
        state.rocketList = rockets;
        state.isLoading = false;
      })
      .addCase(getAPI.rejected, (state, action) => {
        state.isLoading = false;
        state.rocketList = [];

        state.error = action.error.message;
      });
  },
});

export default rocketsSlice.reducer;
