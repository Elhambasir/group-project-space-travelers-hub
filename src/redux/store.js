// eslint-disable-next-line import/no-extraneous-dependencies
import { configureStore } from '@reduxjs/toolkit';
import MissionReducer from './mission/missionSlice';

// eslint-disable-next-line import/prefer-default-export
export const store = configureStore({
  reducer: {
    missions: MissionReducer,
  },
});
