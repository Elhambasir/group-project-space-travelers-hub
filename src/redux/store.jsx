// eslint-disable-next-line import/no-extraneous-dependencies
import { configureStore } from '@reduxjs/toolkit';
import rocketsReducer from './Rockets/rocketsSlice';
import MissionReducer from './mission/missionSlice';

const store = configureStore({
  reducer: {
    rockets: rocketsReducer,
    missions: MissionReducer,
  },
});

export default store;
