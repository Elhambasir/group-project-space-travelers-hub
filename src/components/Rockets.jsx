// eslint-disable-next-line import/no-extraneous-dependencies
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reservation } from '../redux/Rockets/rocketsSlice';

const Rocket = (props) => {
  const dispatch = useDispatch();

  const {
    id, name, disc, image, Reservation, reservationState,
  } = props;
  return (
    <div className='rocketC' key={id}>
      <img className='rocketImg' src={image} alt={name} />
      <div className='contentC'>
      <h4 className='rocektName'>{name}</h4>
      <div className='rocketDesc'>{disc}</div>
      <div type="button" onClick={() => { dispatch(reservation(Reservation)); }}>{reservationState}</div>
      </div>
    </div>
  );
};

Rocket.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  disc: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  Reservation: PropTypes.number.isRequired,
  reservationState: PropTypes.number.isRequired,
};

export default Rocket;
