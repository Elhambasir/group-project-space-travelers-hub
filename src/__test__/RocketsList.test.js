import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import RocketList from '../components/RocketsList';
import store from '../redux/store';
import { reservation } from '../redux/Rockets/rocketsSlice';

describe('Component render testing', () => {
  it('RocketList page renders correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <RocketList />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock('react-redux');

describe('render rockets', () => {
  const rockets = {
    RocketList: [
      {
        id: 1,
        name: 'Falcon 1',
        disc: 'The Falcon 1 was an expendable launch system privately developed and manufactured by SpaceX during 2006-2009. On 28 September 2008, Falcon 1 became the first privately-developed liquid-fuel launch vehicle to go into orbit around the Earth.',
        images: 'https://imgur.com/azYafd8.jpg',
        reserved: true,
      },
      {
        id: 2,
        name: 'Falcon 9',
        disc: 'Falcon 9 is a two-stage rocket designed and manufactured by SpaceX for the reliable and safe transport of satellites and the Dragon spacecraft into orbit.',
        images: 'https://farm4.staticflickr.com/3955/32915197674_eee74d81bb_b.jpg',
        reserved: false,
      },
      {
        id: 3,
        name: 'Falcon Heavy',
        disc: 'With the ability to lift into orbit over 54 metric tons (119,000 lb)--a mass equivalent to a 737 jetliner loaded with passengers, crew, luggage and fuel--Falcon Heavy can lift more than twice the payload of the next closest operational vehicle, the Delta IV Heavy, at one-third the cost.',
        images: 'https://farm5.staticflickr.com/4645/38583830575_3f0f7215e6_b.jpg',
        reserved: false,
      },
      {
        id: 4,
        name: 'Starship',
        disc: 'Starship and Super Heavy Rocket represent a fully reusable transportation system designed to service all Earth orbit needs as well as the Moon and Mars. This two-stage vehicle — composed of the Super Heavy rocket (booster) and Starship (ship) — will eventually replace Falcon 9, Falcon Heavy and Dragon.',
        images: 'https://live.staticflickr.com/65535/48953946911_e60c5bcc5c_b.jpg',
        reserved: false,
      },
    ],
  };
  test('The reservation state of the Falcon 1 should be true', () => {
    store.dispatch(reservation(1));

    expect(rockets.RocketList[0].reserved).toBe(true);
  });

  test('The reservation state of the Falcon 9 should be False', () => {
    store.dispatch(reservation(2));
    store.dispatch(reservation(2));

    expect(rockets.RocketList[1].reserved).toBe(false);
  });
  test('The reservation state of the Starship should be False', () => {
    store.dispatch(reservation(4));
    store.dispatch(reservation(4));
    expect(rockets.RocketList[3].reserved).toBe(false);
  });
});
