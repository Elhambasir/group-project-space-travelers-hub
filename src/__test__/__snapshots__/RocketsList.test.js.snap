// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component render testing RocketList page renders correctly 1`] = `<DocumentFragment />`;
